storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala

```diff
@@ -0,0 +1,364 @@
+    * Call using "a = b * qm"
+    */
+  def multiply(multiplier: Multiplier): RegexPattern
+
+  /**
+    * Equivalent to repeated concatenation. Multiplier consists of a minimum
+    * and a maximum; maximum may be infinite (for Kleene star closure).
+    * Call using "a = b * qm"
+    */
+  def multiply(multiplier: Int): RegexPattern = multiply(new Multiplier(multiplier))
+
+  /**
+    * Return False if there exists a string which the present RegexPattern
+    * can match. Return true if no such string exists. Example of empty
+    * patterns: CharClassPattern()
+    */
+  def isEmpty: Boolean
+
+  def toFsm(alphabet: Option[Set[Char]] = None): Fsm
+
+  //region Operators
+
+  /**
+    * Just calls equivalent method
+    */
+  final def ==(that: RegexPattern): Boolean = equivalent(that)
+
+  /**
+    * Just calls !equivalent method
+    */
+  final def !=(that: RegexPattern): Boolean = !equivalent(that)
+
+  /**
+    * Concatenate a sequence of Regex patterns, regardless of differing classes.
+    */
+  final def +(that: RegexPattern): RegexPattern = concatenate(that)
+
+  /**
+    * Alternate between any two Regex patterns, regardless of differing classes.
+    */
+  final def |(that: RegexPattern): RegexPattern = union(that)
+
+  /**
+    * Intersection between any two Regex patterns.
+    */
+  final def &(that: RegexPattern): RegexPattern = intersection(that)
+
+  /**
+    * Returns RegexPattern.negated
+    */
+  final def unary_~ : RegexPattern = negated
+
+  /**
+    * Returns repeated concatenation of this RegexPattern
+    */
+  final def *(multiplier: Int) = multiply(multiplier)
+  //endregion
+}
+
+/**
+  * Companion object for the CharClassPattern, used only for easy testing
+  */
+object CharClassPattern {
+  def apply(charSet: String) = new CharClassPattern(charSet)
+  def apply(charSet: Seq[Char]) = new CharClassPattern(charSet)
+  def apply(charSet: Set[Char]) = new CharClassPattern(charSet)
+  def apply(charSet: String, negateCharSet : Boolean) = new CharClassPattern(charSet, negateCharSet)
+  def apply(charSet: Seq[Char], negateCharSet : Boolean) = new CharClassPattern(charSet, negateCharSet)
+  def apply(charSet: Set[Char], negateCharSet : Boolean) = new CharClassPattern(charSet, negateCharSet)
+}
+
+/**
+  * A CharClass is basically a Set of symbols. The reason for the
+  * CharClass object instead of using frozenset directly is to allow us to
+  * set a "negated" flag. A CharClass with the negation flag set is assumed
+  * to contain every symbol that is in the alphabet of all symbols but not
+  * explicitly listed inside the Set. e.g. [ \ ^ a ]. This is very handy
+  * if the full alphabet is extremely large, but also requires dedicated
+  * combination functions.
+  */
+final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = false)
+    extends RegexPattern {
+  //chars should consist only of chars
+  require(!charSet.contains(Fsm.anythingElse), "Charset can't contain Fsm.AnythingElse")
```

Comment:
You can check that requirement during object creation in the companion's `apply` method and return `Either` on failure.

Reply:
This (and other exceptions) can't occur in regular use of the RegexPattern class. They're mainly library developer at unit-tests step. Consider them as ICE in the compiler.